src/CanvasRender/index.js

```diff
@@ -36,32 +41,21 @@ export default class CanvasRender extends React.Component {
   }
 
   componentDidMount() {
-    this.handleUpdate(this.props)
-  }
-
-  shouldComponentUpdate(nextProps) {
-    if (this.props.width !== nextProps.width || this.props.height !== nextProps.height) return true
-    return false
+    this.handleUpdate()
   }
 
-  componentDidUpdate() {
-    this.handleUpdate(this.props)
-  }
-
-  componentWillReceiveProps(nextProps) {
-    if (this.props.renderData !== nextProps.renderData) {
-      this.handleUpdate(nextProps)
-    }
-    if (this.props.renderLayers !== nextProps.renderLayers) {
-      this.handleUpdate(nextProps)
+  componentDidUpdate(prevProps) {
+    const dependencies = ['width', 'height', 'renderData', 'renderLayers']
+    if (dependenciesHaveChanged(dependencies, this.props, prevProps)) {
+      this.handleUpdate()
     }
   }
```

Comment:
Oh, interesting. I didn't notice the `shouldComponentUpdate`, so `componentDidUpdate` was only being called in some cases to begin with. I think I understand the original intention now. `width` and `height` were checked in `shouldComponentUpdate` rather than `componentWillReceiveProps` because they need to be reflected in `render`. However, there is still the issue of calling `handleUpdate` up to three times, and it's unsafe to be doing that in `componentWillReceiveProps` anyway.

This solution looks good, but IMO we should simplify even further and just make `CanvasRender` extend `PureComponent`. It seems like the component _should_ respond to changes in its other props (`theme`, `render`, etc.), and the cost of checking a few additional props per render is extremely small (especially when compared to the current method which allocates an array and a function each time).